analysis/priestholy/src/modules/shadowlands/items/FlashConcentration.tsx

```diff
@@ -0,0 +1,243 @@
+      Events.applybuff.by(SELECTED_PLAYER).spell(SPELLS.FLASH_CONCENTRATION),
+      this.onByPlayerFlashConcentrationInitialBuff,
+    );
+    this.addEventListener(
+      Events.removebuff.by(SELECTED_PLAYER).spell(SPELLS.FLASH_CONCENTRATION),
+      this.onByPlayerFlashConcentrationBuffRemoval,
+    );
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(SPELLS.FLASH_HEAL),
+      this.onByPlayerFlashHealCast,
+    );
+    this.addEventListener(
+      Events.heal.by(SELECTED_PLAYER).spell(SPELLS.GREATER_HEAL),
+      this.onByPlayerHealAmount,
+    );
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(SPELLS.GREATER_HEAL),
+      this.onByPlayerHealCast,
+    );
+  }
+
+  /**
+   * Sets _isFullStack to true if there are now 5 stacks, as well as the last refresh
+   * NB: it comes after the Flash Heal Cast
+   * @param event indicate the new number of stacks (only positive increments)
+   */
+  onByPlayerFlashConcentrationBuff(event: ApplyBuffStackEvent) {
+    if (event.stack === 5) {
+      this._isFullStack = true;
+      this._lastFullStackRefresh = event.timestamp;
+    } else {
+      this._isFullStack = false;
+    }
+    this._currentStacks = event.stack;
+  }
+
+  /**
+   * Sets _isFullStack to false.
+   * Note: The only exception is the prepull event, for which we assume that
+   * it means the character has five stacks. It will be corrected if the
+   * first Flash Heal generates a ApplyBuffStack or ApplyBuff event
+   * Note 2: Technically, the only time this function does something is on the first
+   * apply buff if there was no prepull FC.
+   * @param event with info about first stack, or prepull incomplete indication
+   */
+  onByPlayerFlashConcentrationInitialBuff(event: ApplyBuffEvent) {
+    if (event.prepull !== true) {
+      this._isFullStack = false;
+      this._currentStacks = 1;
+    }
+  }
+
+  /**
+   * Increments _totalDrops for our suggestion, since that should ideally not happen
+   */
+  onByPlayerFlashConcentrationBuffRemoval(event: RemoveBuffEvent) {
+    this._totalDrops += 1;
+    this._isFullStack = false;
+    this._currentStacks = 0;
+  }
+
+  /**
+   * when at full stack, it refreshes the last refresh timer,
+   * and counts a Wasteful Cast if it was used earlier than needed
+   * without 2 Surge Of Light Stacks
+   */
+  onByPlayerFlashHealCast(event: CastEvent) {
+    if (this._isFullStack) {
+      if (
+        event.timestamp - this._lastFullStackRefresh < FLASH_CONCENTRATION_DURATION - 5000 &&
+        this.selectedCombatant.getBuffStacks(SPELLS.SURGE_OF_LIGHT_TALENT.id) < 2
+      ) {
+        this._wastefulCasts += 1;
+      }
+      this._lastFullStackRefresh = event.timestamp;
+    }
+  }
+
+  onByPlayerHealAmount(event: HealEvent) {
+    this._totalHeal += event.amount + (event.absorbed || 0);
+    this._gainedHeal += calculateEffectiveHealingStacked(
+      event,
+      FLASH_CONCENTRATION_HEAL_BUFF,
+      this._currentStacks,
+    );
```

Comment:
Wait a sec just realized this is wrong. `calculateEffectiveHealingStacked` is the attribution for the last stack, while I think you want attributions for all the stacks. `calculateEffectiveHealing(event, FLASH_CONCENTRATION_HEAL_BUFF * this._currentStacks)` is what you want here

Reply:
let me check again when I get a chance :) I was not sure myself between both tbh :sweat_smile:

Reply:
Those methods were in dire need of a docs update so I just put up a PR for it. If you've got time could you take your new-person eyes and look at this PR? https://github.com/WoWAnalyzer/WoWAnalyzer/pull/4516  - let me know if you can understand how they work from the new docs.